File: c/loops.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <stdbool.h>
#include <sys/time.h>

#define iterations 10000000
#define sampleSize 23

void simulate();

int main() {
  simulate();
}

void simulate() {
  struct timeval start, end;
  gettimeofday(&start, NULL);
  srand(start.tv_usec);

  int duplicates = 0;
  for (int x=0; x < iterations; x++) {
    int data[365] = {};
    for (int i=0; i < sampleSize; i++) {
      int number = (float) rand() / RAND_MAX * 365;
      // check for number in array
      // bool inArray = false;
      // for (int c=0; c<i; c++) {
      //   if (data[c] == number) inArray = true;
      // }
      if (data[number] == 1) {
        // printf("dupe\n");
        duplicates++;
        break;
      } else {
        data[number] = 1;
      }
    }
  }
  printf("iterations: %d\n", iterations);
  printf("sample-size: %d\n", sampleSize);
  double results = duplicates * 100.0 / iterations;
  printf("percent: %0.2f\n", results);
  gettimeofday(&end, NULL);
  double startTime = start.tv_sec + start.tv_usec / 1000000.0;
  double endTime = end.tv_sec + end.tv_usec / 1000000.0;
  float diff = endTime - startTime;
  printf("seconds: %.3f\n", diff);
}

File: go/loops.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

func main() {
	simulate()
}

func simulate() {
	start := time.Now().UnixNano()
	iterations := 1000000
	sampleSize := 23

	count := 0
	for i := 0; i < iterations; i++ {
		var data [365]int
		for n := 0; n < sampleSize; n++ {
			number := rand.Intn(365)
			if data[number] == 1 {
				count++
				break
			} else {
				data[number] = 1
			}
		}
	}
	fmt.Printf("iterations: %d\n", iterations)
	fmt.Printf("sample-size: %d\n", sampleSize)
	percent := float64(count) / float64(iterations) * 100
	fmt.Printf("percent: %.2f\n", math.Floor(percent*100)/100) // format with printf
	end := time.Now().UnixNano()
	diff := float64(end-start) / 1000 / 1000 / 1000
	fmt.Printf("seconds: %.3f\n", diff)
}
